app/src/main/java/com/chouten/app/data/repository/ModuleRepositoryImpl.kt

```diff
@@ -0,0 +1,261 @@
+package com.chouten.app.data.repository
+
+import android.content.Context
+import android.net.Uri
+import android.provider.DocumentsContract
+import android.webkit.MimeTypeMap
+import androidx.core.net.toUri
+import androidx.documentfile.provider.DocumentFile
+import com.chouten.app.domain.proto.filepathDatastore
+import com.chouten.app.domain.repository.ModuleRepository
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.withContext
+import java.io.IOException
+import java.util.zip.ZipInputStream
+import javax.inject.Inject
+
+class ModuleRepositoryImpl @Inject constructor(
+    val context: Context,
+    val getModuleDir: suspend (Uri) -> Uri,
+) : ModuleRepository {
+
+    /**
+     * Returns a list of all modules in the module folder
+     * A module is viewed as any folder in the module folder
```

Comment:
We could do but it **is** possible for the rename to fail which would mess it up.
I think this approach is fine since the use case will need to parse the module anyway which 
filters out the invalid modules.

Reply:
I completely forgot about the backup folder. It should just be the same logic as copying the zip from an online resource; I'll do it in another PR.